File: crates/r4s_engine/src/advancement.rs
use crate::character::{find_duplicate, normalize_skill_name, slots_full};
use crate::dice::DiceRoll;
use crate::error::EngineError;
use crate::resolve::RollOutcome;
use crate::types::{Character, RoomSettings, Skill};

/// Posibilidad de ganar una habilidad nueva tras una tirada.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AdvancementOption {
    pub source_skill_index: usize,
    pub source_label: String,
    pub new_level: u8,
    /// Todos los dados salieron 6: avance gratis.
    pub natural: bool,
    /// XP necesario: 1 por cada dado que no sea 6.
    pub xp_cost: u32,
    /// XP que se puede gastar en esta tirada.
    pub xp_available: u32,
    /// Hay que elegir qué habilidad reemplazar (o descartar la nueva).
    pub slots_full: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SlotChoice {
    /// Ocupa un slot libre.
    Append,
    /// Reemplaza la habilidad en ese índice (nunca 0).
    Replace(usize),
    /// Renuncia a la habilidad nueva.
    Discard,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AdvancementChoice {
    pub new_skill_name: String,
    pub slot: SlotChoice,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ApplyResult {
    pub character: Character,
    pub xp_gained: u32,
    pub xp_spent: u32,
    pub new_skill: Option<Skill>,
    /// Índice y habilidad reemplazada.
    pub replaced: Option<(usize, Skill)>,
}

fn source_skill(character: &Character, skill_index: usize) -> Result<&Skill, EngineError> {
    character
        .skills
        .get(skill_index)
        .ok_or(EngineError::SkillIndexOutOfRange(skill_index))
}

fn check_roll_matches(skill: &Skill, roll: &DiceRoll) -> Result<(), EngineError> {
    if roll.len() != skill.level as usize {
        return Err(EngineError::DiceCountMismatch {
            expected: skill.level as usize,
            got: roll.len(),
        });
    }
    Ok(())
}

fn xp_available(character: &Character, settings: &RoomSettings, outcome: &RollOutcome) -> u32 {
    if settings.xp_same_roll {
        character.xp + outcome.xp_gained
    } else {
        character.xp
    }
}

/// Calcula si la tirada permite ganar una habilidad nueva.
///
/// Devuelve `None` si no hay avance posible: no todos son 6 y el XP no alcanza,
/// o la habilidad ya está en el nivel máximo.
pub fn advancement_option(
    character: &Character,
    settings: &RoomSettings,
    roll: &DiceRoll,
    outcome: &RollOutcome,
    skill_index: usize,
) -> Result<Option<AdvancementOption>, EngineError> {
    let skill = source_skill(character, skill_index)?;
    check_roll_matches(skill, roll)?;

    if skill.level >= settings.max_dice {
        return Ok(None);
    }

    let xp_cost = roll.non_sixes();
    let available = xp_available(character, settings, outcome);
    if xp_cost > available {
        return Ok(None);
    }

    Ok(Some(AdvancementOption {
        source_skill_index: skill_index,
        source_label: skill.label(),
        new_level: skill.level + 1,
        natural: xp_cost == 0,
        xp_cost,
        xp_available: available,
        slots_full: slots_full(character, settings),
    }))
}

/// Aplica el resultado de una tirada resuelta a la hoja: XP por fallo y,
/// opcionalmente, la habilidad nueva elegida por el jugador.
pub fn apply_roll(
    character: &Character,
    settings: &RoomSettings,
    roll: &DiceRoll,
    outcome: &RollOutcome,
    skill_index: usize,
    choice: Option<&AdvancementChoice>,
) -> Result<ApplyResult, EngineError> {
    let skill = source_skill(character, skill_index)?;
    check_roll_matches(skill, roll)?;

    let mut updated = character.clone();
    updated.xp = character.xp + outcome.xp_gained;

    let choice = match choice {
        Some(c) if c.slot != SlotChoice::Discard => c,
        _ => {
            return Ok(ApplyResult {
                character: updated,
                xp_gained: outcome.xp_gained,
                xp_spent: 0,
                new_skill: None,
                replaced: None,
            });
        }
    };

    let option =
        advancement_option(character, settings, roll, outcome, skill_index)?.ok_or(EngineError::NotEligible)?;

    let name = normalize_skill_name(&choice.new_skill_name)?;
    let new_skill = Skill {
        name,
        level: option.new_level,
        permanent: false,
        derived_from: Some(option.source_label.clone()),
    };

    let mut replaced = None;
    match choice.slot {
        SlotChoice::Append => {
            if option.slots_full {
                return Err(EngineError::SlotsFull);
            }
            if find_duplicate(&updated.skills, &new_skill.name, None) {
                return Err(EngineError::DuplicateSkillName(new_skill.name));
            }
            updated.skills.push(new_skill.clone());
        }
        SlotChoice::Replace(index) => {
            let target = updated
                .skills
                .get(index)
                .ok_or(EngineError::SkillIndexOutOfRange(index))?;
            if target.permanent {
                return Err(EngineError::CannotReplacePermanent);
            }
            if find_duplicate(&updated.skills, &new_skill.name, Some(index)) {
                return Err(EngineError::DuplicateSkillName(new_skill.name));
            }
            let old = std::mem::replace(&mut updated.skills[index], new_skill.clone());
            replaced = Some((index, old));
        }
        SlotChoice::Discard => unreachable!("Discard se maneja arriba"),
    }

    // `advancement_option` ya garantizó que el XP alcanza según `xp_same_roll`.
    updated.xp -= option.xp_cost;

    Ok(ApplyResult {
        character: updated,
        xp_gained: outcome.xp_gained,
        xp_spent: option.xp_cost,
        new_skill: Some(new_skill),
        replaced,
    })
}

File: web/src/engine/advancement.ts
import { findDuplicate, normalizeSkillName, slotsFull } from './character';
import type { DiceRoll } from './dice';
import { EngineError } from './errors';
import type { RollOutcome } from './resolve';
import { skillLabel, slotPrice, type Character, type RoomSettings, type Skill } from './types';

/** Posibilidad de ganar una habilidad nueva tras una tirada. */
export interface AdvancementOption {
  sourceSkillIndex: number;
  sourceLabel: string;
  newLevel: number;
  /** Todos los dados salieron 6: avance gratis. */
  natural: boolean;
  /** XP necesario: 1 por cada dado que no sea 6. */
  xpCost: number;
  /** XP que se puede gastar en esta tirada. */
  xpAvailable: number;
  /** Hay que elegir qué habilidad reemplazar (o descartar la nueva). */
  slotsFull: boolean;
  /** Con los slots llenos y `buySlots` activo: lo que cuesta comprar un slot (2 × nivel nuevo); si no, `null`. */
  slotPrice: number | null;
  /** El XP alcanza para los dados y el slot. */
  canBuySlot: boolean;
}

/** `append` ocupa un slot libre; `replace` sustituye la habilidad en `index` (nunca 0); `buy` compra un slot con XP
 *  y la añade; `discard` renuncia. */
export type SlotChoice = { kind: 'append' } | { kind: 'replace'; index: number } | { kind: 'buy' } | { kind: 'discard' };

export interface AdvancementChoice {
  newSkillName: string;
  slot: SlotChoice;
}

export interface ApplyResult {
  character: Character;
  xpGained: number;
  xpSpent: number;
  newSkill: Skill | null;
  /** Índice y habilidad reemplazada. */
  replaced: { index: number; skill: Skill } | null;
}

function sourceSkill(character: Character, skillIndex: number): Skill {
  const skill = character.skills[skillIndex];
  if (!skill) throw new EngineError({ kind: 'SkillIndexOutOfRange', index: skillIndex });
  return skill;
}

function checkRollMatches(skill: Skill, roll: DiceRoll): void {
  if (roll.length !== skill.level) throw new EngineError({ kind: 'DiceCountMismatch', expected: skill.level, got: roll.length });
}

function xpAvailable(character: Character, settings: RoomSettings, outcome: RollOutcome): number {
  return settings.xpSameRoll ? character.xp + outcome.xpGained : character.xp;
}

/** Calcula si la tirada permite ganar una habilidad nueva.
 *
 *  Devuelve `null` si no hay avance posible: no todos son 6 y el XP no alcanza, o la habilidad ya está en el
 *  nivel máximo. */
export function advancementOption(
  character: Character,
  settings: RoomSettings,
  roll: DiceRoll,
  outcome: RollOutcome,
  skillIndex: number,
): AdvancementOption | null {
  const skill = sourceSkill(character, skillIndex);
  checkRollMatches(skill, roll);

  if (skill.level >= settings.maxDice) return null;

  const xpCost = roll.nonSixes();
  const available = xpAvailable(character, settings, outcome);
  if (xpCost > available) return null;

  const full = slotsFull(character, settings);
  const price = full && settings.buySlots ? slotPrice(skill.level + 1) : null;
  return {
    sourceSkillIndex: skillIndex,
    sourceLabel: skillLabel(skill),
    newLevel: skill.level + 1,
    natural: xpCost === 0,
    xpCost,
    xpAvailable: available,
    slotsFull: full,
    slotPrice: price,
    canBuySlot: price !== null && xpCost + price <= available,
  };
}

function cloneCharacter(c: Character): Character {
  return { ...c, skills: c.skills.map((s) => ({ ...s })) };
}

/** Aplica el resultado de una tirada resuelta a la hoja: XP por fallo y, opcionalmente, la habilidad nueva
 *  elegida por el jugador. */
export function applyRoll(
  character: Character,
  settings: RoomSettings,
  roll: DiceRoll,
  outcome: RollOutcome,
  skillIndex: number,
  choice: AdvancementChoice | null,
): ApplyResult {
  const skill = sourceSkill(character, skillIndex);
  checkRollMatches(skill, roll);

  const updated = cloneCharacter(character);
  updated.xp = character.xp + outcome.xpGained;

  if (!choice || choice.slot.kind === 'discard') {
    return { character: updated, xpGained: outcome.xpGained, xpSpent: 0, newSkill: null, replaced: null };
  }

  const option = advancementOption(character, settings, roll, outcome, skillIndex);
  if (!option) throw new EngineError({ kind: 'NotEligible' });

  const newSkill: Skill = {
    name: normalizeSkillName(choice.newSkillName),
    level: option.newLevel,
    permanent: false,
    derivedFrom: option.sourceLabel,
  };

  let replaced: ApplyResult['replaced'] = null;
  let xpSpent = option.xpCost;
  if (choice.slot.kind === 'append' || choice.slot.kind === 'buy') {
    if (choice.slot.kind === 'buy') {
      if (option.slotPrice === null) throw new EngineError({ kind: 'CannotBuySlot' });
      if (!option.canBuySlot) throw new EngineError({ kind: 'InsufficientXp', needed: option.xpCost + option.slotPrice, available: option.xpAvailable });
      updated.extraSlots += 1;
      xpSpent += option.slotPrice;
    } else if (option.slotsFull) {
      throw new EngineError({ kind: 'SlotsFull' });
    }
    if (findDuplicate(updated.skills, newSkill.name)) throw new EngineError({ kind: 'DuplicateSkillName', name: newSkill.name });
    updated.skills.push({ ...newSkill });
  } else {
    const index = choice.slot.index;
    const target = updated.skills[index];
    if (!target) throw new EngineError({ kind: 'SkillIndexOutOfRange', index });
    if (target.permanent) throw new EngineError({ kind: 'CannotReplacePermanent' });
    if (findDuplicate(updated.skills, newSkill.name, index)) throw new EngineError({ kind: 'DuplicateSkillName', name: newSkill.name });
    updated.skills[index] = { ...newSkill };
    replaced = { index, skill: target };
  }

  // `advancementOption` ya garantizó que el XP alcanza según `xpSameRoll` (y `canBuySlot` con el slot).
  updated.xp -= xpSpent;

  return { character: updated, xpGained: outcome.xpGained, xpSpent, newSkill, replaced };
}
